docs/source/user/stencil.rst

```diff
@@ -0,0 +1,205 @@
+==========================================
+
+In the above example and in most cases, the array indexing in the 
+stencil kernel will exclusively use Integer literals.
+In such cases, the stencil decorator is able to analyze the stencil
+kernel to determine its size.  In the above example, the stencil
+decorator determines that the kernel is 3x3 since indices -1 to 1
+are used for both the first and second dimensions.  Note that
+the stencil decorator also correctly handles non-symmetric and 
+non-square stencil kernels.
+
+Based on the size of the stencil kernel, the stencil decorator is
+able to compute the size of the border in the output array.  If
+applying the kernel to some element of input array would cause
+an index to be out-of-bounds then that element belongs to the border
+of the output array.  In the above example, points -1 and +1 are
+accessed in each dimension and thus the output array has a border
+of size one in all dimensions.
+
+
+Stencil decorator options
+=========================
+
+While the stencil decorator may be augmented in the future to 
+provide additional mechanisms for border handling, at the moment
+the stencil decorator currently supports only one option.
+
+.. _stencil-neighborhood:
+
+``neighborhood``
+----------------
+
+Sometimes it may be inconvenient to write the stencil kernel
+exclusively with Integer literals.  For example, let us say we
+would like to compute the trailing 30-day moving average of a
+time series of data.  One could write 
+``(a[-29] + a[-28] + ... + a[-1] + a[0]) / 30`` but the stencil
+decorator offers a more concise form using the ``neighborhood``
+option::
+
+   @stencil(neighborhood=((-29,0),))
+   def kernel2(a):
+       cumul = 0
+       for i in range(-29,1):
+           cumul += a[i]
+       return cumul / 30
+
+The neighborhood option is a tuple of tuples.  The outer tuple's
+length is equal to the number of dimensions of the input array.
+The inner tuples' lengths are always 2 because
+each element of the outer tuple corresponds to minimum and
+maximum index offsets used in the corresponding dimension.
+
+If a user specifies a neighborhood but then in the kernel 
+accesses elements outside the specified neighborhood, the behavior
+is undefined.
+
+.. _stencil-mode:
+
+``mode``
+--------
+
+The optional mode parameter controls how the border of the output array
+is handled.  Currently, there is only one supported value, ``"constant"``.
+In ``constant`` mode, the stencil kernel is not applied in cases where
+the kernel would access elements outside the valid range of the input
+array.  In such cases, those elements in the output array are assigned
+to a constant value, as specified by the ``cval`` parameter.
+
+``cval``
+--------
+
+The optional cval parameter defaults to zero but can be set to any
+desired value, which is then used for the border of the output array
+if the mode parameter is set to ``constant``.  The cval parameter is 
+ignored in all other modes.
```

Comment:
Should we just require that cval type match the kernel return type?  Or does a specified cval type override the computed kernel type and we automatically insert a coercion to the cval type before assignment?

Reply:
Any idea of the level of effort for them? I imagine the former is easier? I think for performance the assignment would be cheaper on narrower types but it may that a `cval` is set to force the allocation type and saving having to reallocate may be important perhaps later in the user code?

Reply:
Implemented a check for cval type matching the kernel return type.  If the programmer wants a coercion they can just do it theirself in the kernel.